orttraining/orttraining/python/training/onnxblock/_graph_utils.py

```diff
@@ -0,0 +1,144 @@
+import copy
+from onnxruntime.capi._pybind_state import GradientGraphBuilder
+import onnx
+
+from onnxruntime.training import onnxblock
+
+
+def build_gradient_model(model, user_args_requiring_grad, user_args_not_requiring_grad):
+    """Builds the gradient graph on top of the given input forward only graph."""
+
+    # Collect names of parameters that need gradients computed
+    all_args_requiring_gradient = set()
+    # Move all trainable and non trainable initializers to graph inputs.
+    # This allows training to pass in the parameters from outside the graph
+    # so as to share the parameters across multiple sessions.
+    graph_inputs = model.graph.input
+    initializers = []
+    for initializer in model.graph.initializer:
+        if not initializer.name[0].isdigit():
+            # Move only those initializers as inputs that are not local
+            # to the onnx model. i.e. initializers that are model parameters.
+            # These are tpically those initializers without any number prefixed
+            # to their names.
+            graph_inputs.append(
+                onnx.helper.make_tensor_value_info(
+                    initializer.name, initializer.data_type, initializer.dims
+                )
+            )
+            if initializer.name not in user_args_not_requiring_grad:
+                all_args_requiring_gradient.add(initializer.name)
+        else:
+            # All other initializers stay where they were.
+            initializers.append(initializer)
+
+    # Graph and model with initializers as inputs.
+    graph_with_initializers_as_inputs = onnx.helper.make_graph(
+        model.graph.node,
+        "Forward Graph with Initilializers as Inputs",
+        graph_inputs,
+        model.graph.output,
+        initializer=initializers,
+    )
+    grad_model = onnx.helper.make_model(
+        graph_with_initializers_as_inputs,
+        producer_name=onnxblock._producer_name,
+        opset_imports=[onnx.helper.make_opsetid("com.microsoft", 1)]
+        + list(model.opset_import),
+    )
+
+    # Any graph input that requires gradient, should have been already added to
+    # args_requiring_grad.
+    for argument_name in user_args_requiring_grad:
+        all_args_requiring_gradient.add(argument_name)
+
+    # Assumption is that the graph has an output called `loss`.
+    builder = GradientGraphBuilder(
+        grad_model.SerializeToString(), {"loss"}, all_args_requiring_gradient, "loss"
```

Comment:
what if it is called something else

Reply:
Yeah, I think this might be an older version of the file you reviewed. There is no reference to "loss" as such with the new commit. I use a random number generator between 0-1000 and append to the names. We can change this logic going forward if we like.